source/slang/memory_pool.cpp

```diff
@@ -0,0 +1,52 @@
+#include "memory_pool.h"
+
+namespace Slang
+{
+    const size_t kPoolSegmentSize = 4 << 20; // use 4MB segments
+
+    struct MemoryPoolSegment
+    {
+        unsigned char* data = nullptr;
+        size_t allocPtr = 0;
+        MemoryPoolSegment* nextSegment = nullptr;
+    };
+
+    MemoryPool::~MemoryPool()
+    {
+        while (curSegment)
+        {
+            auto nxtSegment = curSegment->nextSegment;
+            free(curSegment->data);
+            delete curSegment;
+            curSegment = nxtSegment;
+        }
+    }
+
+    void newSegment(MemoryPool* pool)
+    {
+        auto seg = new MemoryPoolSegment();
+        seg->nextSegment = pool->curSegment;
+        seg->data = (unsigned char*)malloc(kPoolSegmentSize);
+        pool->curSegment = seg;
```

Comment:
Not a huge deal, but we might want to set things up so that the memory for the `MemoryPoolSegment` comes out of the pool memory too. In particular, this code is using a combination of `new` and `malloc`, when eventually we probably want the `MemoryPool` to bottleneck all its allocations through another allocator (that the user of the Slang library might be able to provide). Of course, we aren't really close to being able to do that any time soon (just because things like `List` and `String` don't have any way to control their allocation).